[WIP] DO NOT MERGE - Initial implementation

ctubbsii opened the pull request:
This is a work in progress. It is not quite ready to merge. I want to make it available for review/feedback before I make any more progress, to make sure it's headed in a good direction.

ctubbsii committed:
Add build tooling for Maven, Travis CI, and Git

ctubbsii committed:
WIP

ctubbsii committed:
WIP

ctubbsii committed:
WIP

ctubbsii committed:
WIP

ctubbsii committed:
WIP

ctubbsii committed:
WIP

keith-turner commented on src/main/java/org/apache/bytes/Bytes.java:
```diff
@@ -0,0 +1,452 @@
+
+  private transient WeakReference<String> utf8String;
+
+  public static final Bytes EMPTY = new Bytes(new byte[0]);
```
Would probably be cleaner to hide this and have a no-arg `of()` method.

ctubbsii replied:
I thought about an `empty()` method, which is the same thing, but a little more intuitive. However, I didn't see much difference between a method and a constant. Are there benefits to the method version over the constant?

keith-turner replied:
I think the main benefit of `of()` is consistency

ctubbsii replied:
I think "of" is the wrong word in this case. I think good fluent interfaces mimic natural language sentences, and an "of" preposition with a blank instead of a direct object to follow would be less "fluent". I think it's probably better to sacrifice consistency by changing the method name in fluent APIs in order to favor readability.

What do you think of `empty()`?

keith-turner replied:
`empty()` is ok w/ me, but I still prefer `of()`.  I have seen both used.  I have seen guava use `of()` in classes like ImmutableMap. Can't remember where I saw `empty()` used.

keith-turner commented on src/main/java/org/apache/bytes/ByteSequence.java:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @since 1.0.0
+ */
+public interface ByteSequence extends Iterable<Byte> {
```
Why make this extend `Iterable<Byte>`?

ctubbsii replied:
I think I was following the CharSequence pattern of being Iterable over Character.

keith-turner commented on src/main/java/org/apache/bytes/AbstractByteSequence.java:
```diff
@@ -0,0 +1,119 @@
+ *
+ * @since 1.0.0
+ */
+abstract class AbstractByteSequence implements ByteSequence {
```
Could provide default implementations of `contentEquals(byte[])` and `compareTo(byte[])`.

keith-turner commented on src/main/java/org/apache/bytes/Bytes.java:
```diff
@@ -0,0 +1,452 @@
+  /**
+   * Returns a byte array containing a copy of the bytes
+   */
+  public byte[] toArray() {
```
Wondering if this method should be in ByteSequence.

ctubbsii replied:
Oh, yeah, that makes sense.

keith-turner commented on src/main/java/org/apache/bytes/ByteSequence.java:
```diff
@@ -0,0 +1,81 @@
+   * @param index the position within the sequence to retrieve
+   * @return the byte at the specified index
+   */
+  byte byteAt(int index);
```
Not sure if this is a good thing to do, but its something I thought of.  Could make index of type long.

keith-turner commented on src/main/java/org/apache/bytes/ByteSequence.java:
```diff
@@ -0,0 +1,81 @@
+   *
+   * @return the length of the sequence
+   */
+  int length();
```
This could possibly be long.  Again, not sure if this is a good idea.

keith-turner commented on src/main/java/org/apache/bytes/Bytes.java:
```diff
@@ -0,0 +1,452 @@
+
+  private int hashCode = 0;
+
+  public Bytes() {
```
If all constructors are package private then, the class is effectively final.  So we can drop final from the class definition.  This allows us the flexibility to subclass Bytes for internal implementations that have very different behaviors.  These internal package private subclasses could only be created by static methods.  This is the approach taken by protobuf.

lapo-luchini committed:
Clarify Javadoc for Bytes.toString()

Update Javadoc for Bytes.toString() method to clarify the strategy
for constructing the Java String is to decode the Bytes with UTF-8.
